lib/plugins/bull-mq-metrics/MetricsCollector.ts

```diff
@@ -0,0 +1,164 @@
+}
+
+type MetricCollectorOptions = Required<Omit<BullMqMetricsPluginOptions, 'redisClient'>>
+
+export class MetricsCollector {
+  private readonly options: MetricCollectorOptions
+  private readonly metrics: Metrics
+  private active = true
+
+  constructor(
+    private readonly redis: Redis,
+    private readonly registry: prometheus.Registry,
+    options: Partial<MetricCollectorOptions>,
+  ) {
+    this.options = {
+      bullMqPrefix: 'bull',
+      metricsPrefix: 'bullmq',
+      excludedQueues: [],
+      collectionIntervalInMs: 5000,
+      histogramBuckets: [20, 50, 150, 400, 1000, 3000, 8000, 22000, 60000, 150000],
+      ...options,
+    }
+
+    this.metrics = this.registerMetrics(this.registry, this.options)
+  }
+
+  async start() {
+    const queueNames = await this.findQueues(this.redis, this.options)
+    await Promise.all(
+      queueNames.map((name) => this.observeQueue(name, this.redis, this.metrics, this.options)),
+    )
+  }
+
+  stop() {
+    this.active = false
+  }
+
+  private registerMetrics(
+    registry: prometheus.Registry,
+    { metricsPrefix, histogramBuckets }: MetricCollectorOptions,
+  ): Metrics {
+    const metrics: Metrics = {
+      completedGauge: new prometheus.Gauge({
+        name: `${metricsPrefix}_jobs_completed`,
+        help: 'Total number of completed jobs',
+        labelNames: ['queue'],
+      }),
+      activeGauge: new prometheus.Gauge({
+        name: `${metricsPrefix}_jobs_active`,
+        help: 'Total number of active jobs (currently being processed)',
+        labelNames: ['queue'],
+      }),
+      failedGauge: new prometheus.Gauge({
+        name: `${metricsPrefix}_jobs_failed`,
+        help: 'Total number of failed jobs',
+        labelNames: ['queue'],
+      }),
+      delayedGauge: new prometheus.Gauge({
+        name: `${metricsPrefix}_jobs_delayed`,
+        help: 'Total number of jobs that will run in the future',
+        labelNames: ['queue'],
+      }),
+      waitingGauge: new prometheus.Gauge({
+        name: `${metricsPrefix}_jobs_waiting`,
+        help: 'Total number of jobs waiting to be processed',
+        labelNames: ['queue'],
+      }),
+      processedDuration: new prometheus.Histogram({
+        name: `${metricsPrefix}_jobs_processed_duration`,
+        help: 'Processing time for completed jobs (processing until completed)',
+        buckets: histogramBuckets,
+        labelNames: ['queue'],
+      }),
+      completedDuration: new prometheus.Histogram({
+        name: `${metricsPrefix}_jobs_completed_duration`,
+        help: 'Completion time for jobs (created until completed)',
+        buckets: histogramBuckets,
+        labelNames: ['queue'],
+      }),
+    }
+
+    for (const metric of Object.values(metrics)) {
+      registry.registerMetric(metric)
+    }
+
+    return metrics
+  }
+
+  private async findQueues(
+    redis: Redis,
+    { bullMqPrefix }: MetricCollectorOptions,
+  ): Promise<string[]> {
+    const scanStream = redis.scanStream({
```

Comment:
there is a much easier (and more performant) way to do this. We register all active queues for future discovery already (this was done for the bull-board).

See https://github.com/lokalise/shared-ts-libs/blob/main/packages/app/background-jobs-common/src/background-job-processor/processors/AbstractBackgroundJobProcessor.ts, "getActiveQueueIds"

Reply:
yeah, I saw that but initially wanted to avoid an extra dependency on something that only works with our BG jobs.

but I can switch to that, I don't mind.

Reply:
`fastify-extras` are conceptually very coupled to our stack specifically (as opposed to, let's say, node-service-template, which is more generic).

It would be great to stay tech-agnostic here, but SCAN iterates over entire DB (in chunk, but still), so it's not exactly the cheapest operation out there :-/

Reply:
>but SCAN iterates over entire DB

yeah, that is true. Alright, I will switch to bg-jobs-common data

Reply:
If you want to reduce the level of coupling, we can extract the queue resolution logic into a separate object, which could be replaced with an alternative implementation via plugin config param.

We can also provide this universal extractor that you wrote as an option for whatever users may not use our libraries.

What do you think?

Reply:
The thing is that bg-jobs-common doesn't really discovers queues, it registers and records their names when the job processor is started.

I'm not sure we can make that abstract enough

Reply:
oh, if you mean - we can just allow to provide a discovery function (and default to our own logic + make generic discoverer available) - that's fine by me

Reply:
yup, that's what I had in mind! sorry for the ambiguity